README.md
# hypeRest

**Est**imation of treatment effects using **hyp**othetical **e**valuations (in **R**), as proposed by Bernheim, Björkegren, Naecker, and Pollmann (2021).

To install this package in `R`, run the following commands:  

```R
library(devtools) 
install_github("michaelpollmann/hypeRest")
```



## Usage

**Note: functions, arguments, and returns may change in future versions. Comments are welcome.**

The package contains a function `simulate_data` to quickly simulate data sets with user-specified dimensions and the data.frame shapes used by the functions in this package.
For low-dimensional specifications, use the function `hyperest`, which requires each variable in the regression to be part of the data.frame.
For high-dimensional specifications, use the function `hypererst_arb`, which uses a formula object to specify the relationship between the outcome and hypothetical evaluations and fixed characteristics.

### Low-dimensional specifications

```r
# simulate a data set with 500 observations, 3 hypothetical variables, 2 fixed characteristics
sims <- simulate_data(n=500, num_h=3, num_x=2)
# extract data
dat_wide <- sims$dat_wide
# estimate effects using hypothetical evaluations; match column names generated inside simulate_data
est_hyp <- hyperest(dat_wide,
                    Y_name="Y",
                    W_name="W",
                    H_names=paste0("H",1:3),
                    H0_names=paste0("H",1:3,"_0"),
                    H1_names=paste0("H",1:3,"_1"),
                    X_names=paste0("X",1:2))
# compare true (in-sample) effect, estimates using hypothetical evaluations, and difference in means
round(c(sims$tau,
        est_hyp$tau,
        mean(sims$dat_wide$Y[dat_wide$W==TRUE]) - mean(sims$dat_wide$Y[dat_wide$W==FALSE])),
      2)

# for additional verification, plot predicted treatment effects against individual-level effects
graphics::plot(est_hyp$tau_i,sims$tau_i)
reg <- stats::lm(tau ~ tau_hat, data = data.frame(tau = sims$tau_i, tau_hat = est_hyp$tau_i))
reg$coefficients
graphics::abline(reg)
```


### High-dimensional specifications

```r
# simulate a data set with 500 observations, 20 hypothetical variables, 10 fixed characteristics,
#   where approximately 3/4 of coefficients are 0
num_h <- 20
num_x <- 10
sims <- simulate_data(n=500, num_h=num_h, num_x=num_x, frac_coeff_0 = 3/4)
# extract data
dat_long <- sims$dat_long
# create variable names and formula using all interactions
all_vars <- c(paste0("H",1:num_h),paste0("X",1:num_x))
formula <- paste("Y","~",paste0("(",paste(all_vars,collapse=" + "),")^2"))
W_name <- "W"

est_hyp_arb <- hyperest_arb(formula=formula,
                            W_name="W",
                            dat_long=dat_long,
                            # don't penalize base hypotheticals
                            dont_penalize = paste0("H",1:num_h),
                            # how to pick lambda in cross-validation
                            s="lambda.min",
                            # which optimizer to use for balanceHD::approx.balance
                            args.approx.balance=list(optimizer = "quadprog"))
# compare true (in-sample) effect, estimates using hypothetical evaluations, and difference in means
round(c(sims$tau,
        est_hyp_arb$tau,
        mean(sims$dat_wide$Y[sims$dat_wide$W==TRUE])
          - mean(sims$dat_wide$Y[sims$dat_wide$W==FALSE])),
      2)

# for additional verification, plot predicted treatment effects against individual-level effects
graphics::plot(est_hyp_arb$tau_i,sims$tau_i)
reg <- stats::lm(tau ~ tau_hat, data = data.frame(tau = sims$tau_i, tau_hat = est_hyp_arb$tau_i))
reg$coefficients
graphics::abline(reg)
```



## Brief Description

Our method proceeds in 2 steps.

1. Estimate the relationship between outcome `Y` and hypothetical evaluations and fixed characteristics `H,X`:
        `Y ~ mu(H, X, theta)`
    where `H` are hypothetical evaluations corresponding to the realized treatment state, and `mu` is the specification, for instance linear or including interaction terms.
2. Predict treatment effects for both treatment states:
        `tau = mu(H(1), X, theta) - mu(H(0), X, theta)`

For high-dimensional specifications with LASSO penalty, `hyperest_arb`, an *approximate residual balancing* (Athey et al., 2018) term is added. Approximate residual balancing is implemented in the `R` package [balanceHD](https://github.com/swager/balanceHD).


## Reference

B. Douglas Bernheim, Daniel Björkegren, Jeffrey Naecker, and Michael Pollmann. **Causal Inference from Hypothetical Evaluations**. 2021.

Susan Athey, Guido W. Imbens, and Stefan Wager. **Approximate Residual Balancing: De-Biased Inference of Average Treatment Effects in High Dimensions**. *Journal of the Royal Statistical Society: Series B (Statistical Methodology)* 80, no. 4 (2018): 597-623. [[arXiv](https://arxiv.org/pdf/1604.07125.pdf)]


NEWS.md
# Version 0.1.0

- initial submit
- `simulate_data` to create example data
- `hyperest` for low-dimensional specifications
- `hyperest_arb` for high-dimensional specifications with LASSO penalty and approximate residual balancing


R/hyperest.R
#' Estimate treatment effects using hypothetical evaluations
#'
#' This function estimates treatment effects and standard errors using hypothetical evaluations.
#'
#' @details
#' Step 1 estimates a linear regression:
#'   \deqn{Y ~ H + X}
#' where the outcome variable Y is given in the input string Y_name,
#' the hypothetical evaluations for the realized treatment state are H given in the input string vector of column names H_names,
#' and optional fixed characteristics (not varying by treatment state) are given by the input string vector of column names X_names.
#'
#' Step 2 estimates treatment effects as the difference in predictions for the treated and control state:
#'   \deqn{\tau = (H(1) - H(0)) \beta}
#' where H(1) are the hypothetical evaluations for the treated state given by the input string column names H1_names,
#' H(0) are the hypothetical evaluations for the control state given by the input column names H0_names,
#' and \eqn{\beta} are the estimated coefficients on H in the regression of step 1.
#'
#' The function estimates the ATE, ATT, ATC, and their standard errors.
#' It also predicts individual level treatment effects, but no standard errors are provided
#' because theoretical properties are only available for average effects within groups (that grow with sample size).
#'
#' @references
#'
#' B. Douglas Bernheim, Daniel Björkegren, Jeffrey Naecker, and Michael Pollmann. "Causal Inference from Hypothetical Evaluations." 2021.
#'
#'
#' @param dat_wide dataframe or equivalent, holding data on separate (assumed to be independent) observations as rows,
#'           with column corresponding to different variables.
#' @param Y_name string, the name of the column of dat_wide that holds the outcome data.
#' @param W_name string, the name of the column of dat_wide that holds the treatment indicator. It is only used to estimate the ATT and ATC.
#' @param H_names string vector, the names of columns holding hypothetical evaluations for the realized treatment state.
#' @param H0_names string vector, the names of columns holding hypothetical evaluations in the absence of treatment.
#'           IMPORTANT: order of variables must be the same as in H_names.
#' @param H1_names string vector, the names of columns holding hypothetical evaluations in the presence of treatment.
#'           IMPORTANT: order of variables must be the same as in H_names.
#' @param X_names (optional) string vector, the names of columns holding fixed characteristics that do not vary by treatment state.
#' @param ... (optional) parameters passed to vcovHC of the sandwich package,
#'           determining how to calculate heteroskedasticity-robust standard errors for the linear regression in step 1.
#'           If no parameters are supplied, the defaults of sandwich::vcovHC are used.
#'
#' @return list:
#'         \item{tau}{estimate of the average treatment effect}
#'         \item{tau_att}{estimate of the average treatment effect on the treated (ATT)}
#'         \item{tau_atc}{estimate of the average treatment effect on the control (ATC)}
#'         \item{se}{standard error of the ATE estimator}
#'         \item{se_att}{standard error of the ATT estimator}
#'         \item{se_atc}{standard error of the ATC estimator}
#'         \item{se_naive_only_s1}{standard error of the ATE estimator taking into account only the step 1 regression variance (holding fixed hypothetical evaluations in step 2)}
#'         \item{se_naive_only_s2}{standard error of the ATE estimator taking into account only the step 2 variance due to sampling of hypothetical evaluations (holding fixed the step 1 coefficient estimates)}
#'         \item{reg}{`lm` regression object of step 1}
#'         \item{tau_i}{predictions of individual-level treatment effects}
#' @examples
#' # simulate a data set with 500 observations, 3 hypothetical variables, 2 fixed characteristics
#' sims <- simulate_data(n=500, num_h=3, num_x=2)
#' # extract data
#' dat_wide <- sims$dat_wide
#' # estimate effects using hypothetical evaluations; match column names generated inside simulate_data
#' est_hyp <- hyperest(dat_wide,
#'                     Y_name="Y",
#'                     W_name="W",
#'                     H_names=paste0("H",1:3),
#'                     H0_names=paste0("H",1:3,"_0"),
#'                     H1_names=paste0("H",1:3,"_1"),
#'                     X_names=paste0("X",1:2))
#' # compare true (in-sample) effect, estimates using hypothetical evaluations, and difference in means
#' round(c(sims$tau,
#'         est_hyp$tau,
#'         mean(sims$dat_wide$Y[dat_wide$W==TRUE]) - mean(sims$dat_wide$Y[dat_wide$W==FALSE])),
#'       2)
#'
#' # for additional verification, plot predicted treatment effects against individual-level effects
#' graphics::plot(est_hyp$tau_i,sims$tau_i)
#' reg <- stats::lm(tau ~ tau_hat, data = data.frame(tau = sims$tau_i, tau_hat = est_hyp$tau_i))
#' reg$coefficients
#' graphics::abline(reg)
#'
#' @export
hyperest <- function(dat_wide
                     , Y_name
                     , W_name
                     , H_names
                     , H0_names
                     , H1_names
                     , X_names=NULL
                     , ...
) {

  # number of observations
  N <- nrow(dat_wide)

  # create formula for linear regression
  # regress Y on S
  formula_string <- paste(Y_name,
                          paste(H_names, collapse=' + '),
                          sep=' ~ ')
  # control for X if specified
  if (!is.null(X_names)) {
    formula_string <- paste(formula_string,
                            paste(X_names, collapse=' + '),
                            sep=' + ')
  }

  formula <- stats::as.formula(formula_string)

  reg <- stats::lm(formula, data=dat_wide)
  # summary(reg)


  # get the regression coefficients of S
  gamma <- as.vector(reg$coefficients[H_names])
  # get the difference in subjective responses
  S1mS0 <- as.matrix(dat_wide[,H1_names] - dat_wide[,H0_names],ncol=length(H_names))
  # individual predictions
  tau_i <- S1mS0 %*% gamma
  # difference in conditional expectations
  tau <- mean(tau_i)
  tau_att <- mean(tau_i[dat_wide[[W_name]]==TRUE])
  tau_atc <- mean(tau_i[dat_wide[[W_name]]==FALSE])


  # estimate standard error

  # expectation of squared first moment
  g <- as.vector(tau - S1mS0 %*% gamma)
  Eg2 <-  mean(g^2)

  # expectation of derivative of first moment w.r.t. gamma (coefficients of S)
  Edgdgamma <- as.vector(colMeans(S1mS0))

  # robust standard errors (covariance matrix) of gamma hat
  Vgamma <- N*sandwich::vcovHC(reg,...)[2:(length(H_names)+1),2:(length(H_names)+1)]

  # expectation of derivative of first moment w.r.t. all regression coefficients
  Edgdtheta <- c(Edgdgamma, 0, rep(0,times=length(X_names)))  # zeros correspond to constant term and X

  # expectation of derivative of second moments (OLS) w.r.t. all regression coefficients
  if (!is.null(X_names)) {
    x <- as.matrix(cbind(dat_wide[,H_names],rep(1,N), dat_wide[,X_names]))
  }
  else {
    x <- as.matrix(cbind(dat_wide[,H_names],rep(1,N)))
  }
  Edmdtheta <- - t(x)%*%x / N

  # expectation of cross-term of moments
  e <- reg$residuals
  m <- x * e
  Egm <- colMeans(g * m)

  # putting it all together
  V_tau <- Eg2 + Edgdgamma %*% Vgamma %*% Edgdgamma - 2 * Edgdtheta %*% solve(Edmdtheta) %*% Egm

  # s.e. for ATT
  p <- mean(dat_wide[[W_name]])
  Eg2W <-  mean((dat_wide[[W_name]]*g)^2)
  EgmW <- colMeans(dat_wide[[W_name]] * g * m)
  EdgdgammaW <- as.vector(colSums(as.matrix(S1mS0[dat_wide[[W_name]]==TRUE,],ncol=length(H_names)))/N)
  EdgdthetaW <- c(EdgdgammaW, 0, rep(0,times=length(X_names)))  # zeros correspond to constant term and X
  V_tau_att <- (Eg2W + EdgdgammaW %*% Vgamma %*% EdgdgammaW
                   - 2 * EdgdthetaW %*% solve(Edmdtheta) %*% EgmW)/(p^2)
  # s.e. for ATC
  p <- mean(dat_wide[[W_name]])
  Eg2W0 <-  mean(((1-dat_wide[[W_name]])*g)^2)
  EgmW0 <- colMeans((1-dat_wide[[W_name]]) * g * m)
  EdgdgammaW0 <- as.vector(colSums(as.matrix(S1mS0[dat_wide[[W_name]]==FALSE,],ncol=length(H_names)))/N)
  EdgdthetaW0 <- c(EdgdgammaW0, 0, rep(0,times=length(X_names)))  # zeros correspond to constant term and X
  V_tau_atc <- (Eg2W0 + EdgdgammaW0 %*% Vgamma %*% EdgdgammaW0
                   - 2 * EdgdthetaW0 %*% solve(Edmdtheta) %*% EgmW0)/((1-p)^2)


  # variance of naive estimator
  tau_se_naive_only_s1 <- as.vector(sqrt(as.vector(colMeans(S1mS0)) %*% Vgamma %*% as.vector(colMeans(S1mS0)) / N))
  tau_se_naive_only_s2 <- as.vector(sqrt(t(gamma)%*%stats::var(S1mS0)%*%gamma / N))

  out <- list(tau = as.numeric(tau)
              , tau_att = as.numeric(tau_att)
              , tau_atc = as.numeric(tau_atc)
              , se = as.numeric(sqrt(diag((V_tau/N))))
              , se_att = as.numeric(sqrt(diag((V_tau_att/N))))
              , se_atc = as.numeric(sqrt(diag((V_tau_atc/N))))
              , se_naive_only_s1 = as.numeric(tau_se_naive_only_s1)
              , se_naive_only_s2 = as.numeric(tau_se_naive_only_s2)
              , reg = reg
              , tau_i = as.numeric(tau_i))
  return(out)
}


R/simulate_data.R
#' Simulate a data set of hypothetical evaluations
#'
#' @details
#' This function simulates data that can be used to illustrate to illustrate the estimation methods included in hypeRest package.
#' The data generating process is **not** calibrated to resemble any real data set.
#' The data are simulated such that the simple difference in means estimator is (severely) biased.
#'
#' @param n integer, sample size, must be large enough to allow regression on hypotheticals and fixed characteristics.
#' @param num_h integer, number of hypothetical evaluation variables, must be at least 1.
#' @param num_x integer, number of fixed characteristics, can be 0.
#' @param frac_coeff_0 numeric, between 0 and 1, probability of each coefficient being 0.
#'
#' @returns list:
#'          \item{dat_wide}{data.frame with `n` rows and `4+3*num_h+num_x` columns.
#'                            Column Y holds observed outcomes.
#'                            Column W holds realized treatment state.
#'                            Columns H1 through H<num_h> hold hypothetical evaluations corresponding to the realized treatment state.
#'                            Columns H1_0 through H<num_h>_0 hold hypothetical evaluations corresponding to the control state.
#'                            Columns H1_1 through H<num_h>_1 hold hypothetical evaluations corresponding to the treated state.
#'                            Columns X1 through X<num_x> hold fixed characteristics.
#'                            Columns Y1 and Y0 hold potential outcomes (under treatment and control) that are not typically available in real data sets}
#'          \item{dat_long}{data.frame with `2*n` rows and `2+num_h+num_x` columns.
#'                            The first `n` rows contain data under the control state of each observation,
#'                            the last `n` rows contain data under the treated state of each observation;
#'                            observations are in the same order in each half.
#'                            Column Y holds the outcome (if observed under the treatment state, otherwise NA).
#'                            Column W indicates the treatment state for which the row contains data.
#'                            Columns H1 through H<num_h> hold hypothetical evaluations for the treatment state of the row.
#'                            Columns X1 through X<num_x> hold fixed characteristics.}
#'          \item{tau_i}{vector of length `n`. Individual-level treatment effects (in the same order as the observations in `dat_wide` and `dat_long`.}
#'          \item{tau}{numeric. The true in-sample average treatment effect.}
#'
#' @examples
#' # simulate a data set with 500 observations, 3 hypothetical variables, 2 fixed characteristics
#' sims <- simulate_data(n=500, num_h=3, num_x=2)
#' # extract data
#' dat_wide <- sims$dat_wide
#' # estimate effects using hypothetical evaluations; match column names generated inside simulate_data
#' est_hyp <- hyperest(sims$dat_wide,
#'                     Y_name="Y",
#'                     W_name="W",
#'                     H_names=paste0("H",1:3),
#'                     H0_names=paste0("H",1:3,"_0"),
#'                     H1_names=paste0("H",1:3,"_1"),
#'                     X_names=paste0("X",1:2))
#' # compare true (in-sample) effect, estimates using hypothetical evaluations, and difference in means
#' round(c(sims$tau,
#'         est_hyp$tau,
#'         mean(dat_wide$Y[dat_wide$Y==TRUE]) - mean(dat_wide$Y[dat_wide$Y==FALSE])),
#'       2)
#'
#' @export
simulate_data <- function(n=300, num_h = 5, num_x = 5, frac_coeff_0=0) {

  if (!is.numeric(n)) { stop("n must be an integer") }
  if (!(n==as.integer(n))) { stop("n must be an integer") }

  if (!is.numeric(num_h)) { stop("num_h must be an integer >= 1") }
  if (!(num_h==as.integer(num_h))) { stop("num_h must be an integer >= 1") }
  if (num_h < 1) { stop("num_h must be an integer >= 1") }

  if (!is.numeric(num_x)) { stop("num_x must be an integer >= 0") }
  if (!(num_x==as.integer(num_x))) { stop("num_x must be an integer >= 0") }
  if (num_x < 0) { stop("num_x must be an integer >= 0") }

  if (n < num_h + num_x + 2) { stop("n must be an integer >= num_h + num_x + 2") }

  if (!is.numeric(frac_coeff_0)) { stop("frac_coeff_0 must be a numeric >= 0 and <= 1") }
  if ((frac_coeff_0 < 0) | (frac_coeff_0>1)) { stop("frac_coeff_0 must be a numeric >= 0 and <= 1") }

  # simulate hypothetical evaluations
  Sigma_h <- matrix(0.25,nrow=num_h,ncol=num_h)
  diag(Sigma_h) <- 1
  H0 <- MASS::mvrnorm(n=n, mu=rep.int(0,num_h), Sigma=Sigma_h)
  H1 <- 0.8*H0 + 0.25 + sqrt(1-1*0.8^2) * MASS::mvrnorm(n=n, mu=rep.int(0,num_h), Sigma=Sigma_h)
  # simulate fixed characteristics
  Sigma_x <- matrix(0.25,nrow=num_x,ncol=num_x)
  diag(Sigma_x) <- 1
  X <- (H0 %*% matrix(stats::runif(n=num_h*num_x),nrow=num_h,ncol=num_x)
        + MASS::mvrnorm(n=n, mu=rep.int(0,num_x), Sigma=Sigma_x))

  # treatment assignment
  gamma_h <- stats::runif(n=num_h, min=0.1) * (stats::runif(n=num_h) > frac_coeff_0)
  gamma_x <- stats::runif(n=num_x, min=0.1) * (stats::runif(n=num_x) > frac_coeff_0)
  W <- scale((H1 - H0) %*% gamma_h + X %*% gamma_x) + stats::rnorm(n=n, sd=2)
  W <- W >= stats::median(W)

  # hypothetical evaluation of observed treatment state
  H <- apply(H1,2,function(h) h*W) + apply(H0,2,function(h) h*(1-W))

  # outcome
  beta_h <- 0.2 * gamma_h + stats::runif(n=num_h) * (stats::runif(n=num_h) > frac_coeff_0)
  beta_x <- 0.2 * gamma_x + stats::runif(n=num_x) * (stats::runif(n=num_x) > frac_coeff_0)
  Y0 <- H0 %*% beta_h + X %*% beta_x + stats::rnorm(n=n,sd=1)
  Y1 <- H1 %*% beta_h + X %*% beta_x + stats::rnorm(n=n,sd=1)
  Y <- W*Y1 + (1-W)*Y0

  dat_wide <- as.data.frame(cbind(Y,W,H,H0,H1,X,Y0,Y1))
  col_names <- c("Y","W",paste0("H",1:num_h),paste0("H",1:num_h,"_0"),paste0("H",1:num_h,"_1"))
  if (num_x > 0) {
    col_names <- c(col_names,paste0("X",1:num_x))
  }
  col_names <- c(col_names, "Y0", "Y1")
  colnames(dat_wide) <- col_names

  dat_long <- as.data.frame(rbind(cbind(ifelse(W==FALSE,Y,NA),FALSE,H0,X),
                                  cbind(ifelse(W==TRUE,Y,NA),TRUE,H1,X)))
  col_names <- c("Y","W",paste0("H",1:num_h))
  if (num_x > 0) {
    col_names <- c(col_names,paste0("X",1:num_x))
  }
  colnames(dat_long) <- col_names

  return(list(dat_wide=dat_wide,
              dat_long=dat_long,
              tau_i = Y1 - Y0,
              tau = mean(Y1 - Y0)))
}


R/hyperest_arb.R
#' Estimate treatment effects using hypothetical evaluations in high-dimensional specifications
#'
#' This function estimates treatment effects and standard errors
#' using hypothetical evaluations in high-dimensional specifications.
#' It uses cross-validated LASSO (glmnet::cv.glmnet) to estimate high-dimensional step 1 regression
#' and approximate residual balancing based on Athey, Imbens, and Wager (2018, implemented in balanceHD).
#'
#' @details
#' Step 1 estimates a LASSO regression based on the formula input, with penalty parameter chosen using cross-validation.
#'
#' Step 1b calculates an approximate residual balancing correction using balanceHD::approx.balance (Athey, Imbens, and Wager, 2018).
#'
#' Step 2 estimates treatment effects as the difference in predictions for the treated and control state:
#'   \deqn{\tau = (\mu(H(1),X) - \mu(H(0),X)) + r}
#' where \eqn{\mu} is the prediction function estimated in step 1 and \eqn{r} is the residual balancing term estimated in step 1b.
#'
#' The function currently estimates the ATE the standard errors of the estimator.
#' It also predicts individual level treatment effects, but no standard errors are provided
#' because theoretical properties are only available for average effects within groups (that grow with sample size).
#'
#'
#' @references
#'
#' B. Douglas Bernheim, Daniel Björkegren, Jeffrey Naecker, and Michael Pollmann. "Causal Inference from Hypothetical Evaluations." 2021.
#'
#' Susan Athey, Guido W. Imbens, and Stefan Wager. "Approximate Residual Balancing: De-Biased Inference of Average Treatment Effects in High Dimensions." Journal of the Royal Statistical Society: Series B (Statistical Methodology) 80, no. 4 (2018): 597-623.
#'
#'
#' @param formula formula (or string interpretable as formula),
#'                 specifies the regression of outcome on hypothetical evaluations and fixed characteristics.
#' @param W_name string, the name of the column of dat_long that holds the treatment indicator.
#' @param dat_long data.frame or equivalent. **Should have 2 rows for each observation**:
#'                 one for data corresponding to the treated state (`[[W_name]] == TRUE`)
#'                 and one for data corresponding to the control state (`[[W_name]] == FALSE`).
#'                 Each row contains outcome data (`NA` if the observations is unobserved under that treatment state),
#'                 the column named `W_name` indicating whether the row corresponds to treatment or control,
#'                 hypothetical evaluations for the treatment state of the row,
#'                 fixed characteristics (should appear in the rows for both treatment states).
#'                 **NOTE**: to interpret individual-level effects, the order of observations
#'                 within `[[W_name]] == TRUE` should be the same as the order of observations within `[[W_name]] == FALSE`.
#'                 For the average treatment effect estimate and standard error, the order of observations is irrelevant.
#' @param dont_penalize (optional) string vector, *variable names* (from the formula, see example) that should not be penalized in the step 1 LASSO regression.
#' @param args.approx.balance (optional) list, arguments passed to balanceHD::approximate.balance, see example for setting the optimization routine
#' @param ... (optional) parameters passed to `glmnet::cv.glmnet` and to `predict` such as number of folds, and criterion for selecting the penalty parameter.
#'
#' @return list of three elements:
#'         \item{tau}{estimate of the average treatment effect (ATE)}
#'         \item{se}{standard error of the ATE estimator}
#'         \item{tau_i}{predictions of individual-level treatment effects (see comment on input argument dat_long to ensure correct order of rows)}
#'
#' @examples
#' # simulate a data set with 500 observations, 20 hypothetical variables, 10 fixed characteristics,
#' #   where approximately 3/4 of coefficients are 0
#' num_h <- 20
#' num_x <- 10
#' sims <- simulate_data(n=500, num_h=num_h, num_x=num_x, frac_coeff_0 = 3/4)
#' # extract data
#' dat_long <- sims$dat_long
#' # create variable names and formula using all interactions
#' all_vars <- c(paste0("H",1:num_h),paste0("X",1:num_x))
#' formula <- paste("Y","~",paste0("(",paste(all_vars,collapse=" + "),")^2"))
#' W_name <- "W"
#'
#' est_hyp_arb <- hyperest_arb(formula=formula,
#'                             W_name="W",
#'                             dat_long=dat_long,
#'                             # don't penalize base hypotheticals
#'                             dont_penalize = paste0("H",1:num_h),
#'                             # how to pick lambda in cross-validation
#'                             s="lambda.min",
#'                             # which optimizer to use for balanceHD::approx.balance
#'                             args.approx.balance=list(optimizer = "quadprog"))
#' # compare true (in-sample) effect, estimates using hypothetical evaluations, and difference in means
#' round(c(sims$tau,
#'         est_hyp_arb$tau,
#'         mean(sims$dat_wide$Y[sims$dat_wide$W==TRUE])
#'           - mean(sims$dat_wide$Y[sims$dat_wide$W==FALSE])),
#'       2)
#'
#' # for additional verification, plot predicted treatment effects against individual-level effects
#' graphics::plot(est_hyp_arb$tau_i,sims$tau_i)
#' reg <- stats::lm(tau ~ tau_hat, data = data.frame(tau = sims$tau_i, tau_hat = est_hyp_arb$tau_i))
#' reg$coefficients
#' graphics::abline(reg)
#'
#' @export
hyperest_arb <- function(formula
                         , W_name
                         , dat_long
                         , dont_penalize=NULL
                         , args.approx.balance=list()
                         , ...
) {
  # regression model to get outcome and covariates for observed treatment state
  reg_obs <- stats::lm(formula=formula, data=dat_long, y=TRUE, x=TRUE, model=FALSE, qr=FALSE)

  # regression model to get (transformations of) covariates for both treatment states
  # use a dependent variable that exists irrespective of observed treatment
  f_reg_all <- paste(W_name,"~",paste(labels(stats::terms(stats::as.formula(formula))),collapse=" + "))
  reg_all <- stats::lm(formula=f_reg_all, data=dat_long, x=TRUE, model=FALSE, qr=FALSE)

  # residual balancing weights
  # balance target
  mean_x1 <- colMeans(reg_all$x[dat_long[[W_name]]==TRUE,])
  mean_x0 <- colMeans(reg_all$x[dat_long[[W_name]]==FALSE,])
  # balance
  gamma1 <- do.call(balanceHD::approx.balance,
                    c(list(M = reg_obs$x,
                           balance.target = mean_x1),
                      args.approx.balance))
  gamma0 <- do.call(balanceHD::approx.balance,
                    c(list(M = reg_obs$x,
                           balance.target = mean_x0),
                      args.approx.balance))

  # cross-validated LASSO
  reg_lasso <- glmnet::cv.glmnet(x=reg_obs$x, y=reg_obs$y,
                                 penalty.factor = !(colnames(reg_obs$x) %in% dont_penalize))
  # predict outcomes
  Y1_hat <- stats::predict(reg_lasso, newx = reg_all$x[dat_long[[W_name]]==TRUE,], ...)
  Y0_hat <- stats::predict(reg_lasso, newx = reg_all$x[dat_long[[W_name]]==FALSE,], ...)

  # then re-weight the residuals
  Y_hat <- stats::predict(reg_lasso, newx = reg_obs$x, ...)
  w_resid <- sum((gamma1-gamma0)*(reg_obs$y - Y_hat))

  # estimate effects and standard error
  tau_i <- Y1_hat - Y0_hat + w_resid
  tau <- mean(tau_i)
  se <- sqrt(sum((gamma1-gamma0)^2*(reg_obs$y - Y_hat)^2) *
               # degrees of freedom correction
               length(gamma1) / max(1, (length(gamma1) - sum(stats::coef(reg_lasso,...) != 0))))


  out <- list(tau = as.numeric(tau),
              se = as.numeric(se),
              tau_i = as.numeric(tau_i))

  return(out)
}
